README.md
# NAME

List::MoreUtils::XS - Provide compiled List::MoreUtils functions

# SYNOPSIS

    use List::Moreutils::XS ();
    use List::MoreUtils ':all';

    my @procs = get_process_stats->fetchall_array;
    # sort by ppid, then pid
    qsort { $a->[3] <=> $b->[3] or $a->[2] <=> $b->[2] } @procs;
    while( @procs ) {
        my $proc = shift @procs;
        my @children = equal_range { $_->[3] <=> $proc->[2] } @procs;
    }

    my @left = qw(this is a test);
    my @right = qw(this is also a test);
    my %rlinfo = listcmp @left, @right;

    # on unsorted
    my $i = firstidx { $_ eq 'yeah' } @foo;
    # on sorted - always first, but might not be 'yeah'
    my $j = lower_bound { $_ cmp 'yeah' } @bar;
    # on sorted - any of occurances, is surely 'yeah'
    my $k = bsearchidx { $_ cmp 'yeah' } @bar;

# SEE ALSO

[List::Util](https://metacpan.org/pod/List::Util), [List::AllUtils](https://metacpan.org/pod/List::AllUtils)

# AUTHOR

Jens Rehsack &lt;rehsack AT cpan.org>

Adam Kennedy &lt;adamk@cpan.org>

Tassilo von Parseval &lt;tassilo.von.parseval@rwth-aachen.de>

# COPYRIGHT AND LICENSE

Some parts copyright 2011 Aaron Crane.

Copyright 2004 - 2010 by Tassilo von Parseval

Copyright 2013 - 2017 by Jens Rehsack

All code added with 0.417 or later is licensed under the Apache License,
Version 2.0 (the "License"); you may not use this file except in compliance
with the License. You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

All code until 0.416 is licensed under the same terms as Perl itself,
either Perl version 5.8.4 or, at your option, any later version of
Perl 5 you may have available.


Changes
Revision history for Perl extension List-MoreUtils-XS

0.429_002	2020-10-05
    - fix parts of RT#132043: listcmp misbehave in XS implementation
      (thanks to Henrik Pauli for reporting)

0.429_001	2020-10-02
    - fix parts of RT#123989: more $a/$b/$_ refcounting bugs
      since some bugs are simply reported wrong, dealing with them breaks
      more (including running code) than it solves. Way to heavy to feel
      better ...
    - introduce functions slide and slideatatime (wished by SCHWERN)
    - toolchain fixes
    - fix RT#133128 - LANG=nb_NO.utf8 related str tests fails (thanks, SJN
      for reporting and provide a reasonable fix)
    - fix issue#7 (Typo), thanks AMBS for reporting

0.428	2017-12-19
    - release 0.427_002 without further changes

0.427_002   2017-12-14
    - Update ppport.h to 3.37 for improved blead support
    - Fix RT#123869 - context arg to croak() reported by Zefram
    - fix RT#123870 - one() returns true on empty list reported by Andy Lester
    - pamper RT#123868 - $a/$b/$_ refcounting bugs reported by Zefram

0.427_001   2017-12-12
    - Fix RT#123613 - build fails on perl >=5.27.4 with -DDEBUGGING,
      thanks for reporting and the suggested patch goes to Andreas
      Koenig and Tony Cook

0.426	2017-10-02
    - release 0.425_001 as 0.426

0.425_001	2017-09-27
    - Fix broken format in part (fixes issues/4, thanks Andreas Schwab
      for reporting and testing the fix)
    - Fix gcc 4.7 sequence point warning (fixes issues/5, thanks Andreas
      Schwab for reporting and testing the fix)
    - Fix incorrect padname resolving for perl > 5.21.6 (Fixes RT#122883,
      thanks Alexander Bluhm <alexander.bluhm@gmx.net> for reporting).
    - Fix compiling issue on CentOS 4 & CentOS 5 (thanks Alex Karelas
      <alex.karelas@gmail.com> for reporting and testing the fix)

0.423	2017-08-19
    - Fix dealing with lists with one element on bremove/binsert.
      Thanks shawnlaffan for reporting this via issues#2
    - add support for compilers before C99
    - fix some 32-bit compiler warnings
    - add support for compilers without statement expression feature

0.422	2017-08-15
    - Rename 'occurances' into 'occurrences' as requested in RT#122806 by
      Paul Howarth as a slightly late side-note in RT#91991.
    - Add DESCRIPTION to Pod clarifying the role of List::MoreUtils::XS
    - improve Makefile.PL regarding some build artifacts

0.421	2017-08-14
    - correct license in META thanks to Chase Whitener (RT#122702)
    - fix issues with -DPERL_IMPLICIT_SYS on Windows with Strawberry-Perl

0.420_002	2017-08-03
    - improve configure toolchain to use Config::AutoConf 0.315
    - speed up some inner loops by hinting the expected result
    - fix mind screwed up issue in upper_bound and reduce elements
      visited in equal_range

0.420_001	2017-06-28
    - fix a lot of potential memory leaks when callbacks throw exceptions
    - add some new functions:
      * qsort
      * binsert
      * bremove
      * listcmp
      * arrayify (RT#17230)
      * samples (RT#77562)
      * minmaxstr (RT#106401)
      * lower_bound
      * upper_bound
      * equal_range
      * frequencies
      * occurances
      * mode (RT#91991)
      * zip6 (RT#42921)
      * reduce_0
      * reduce_1
      * reduce_u
    - improve tests
    - make List::MoreUtils::XS independent from List::MoreUtils
      Note that List::MoreUtils::XS doesn't guarantee API stability - this
      feature is only provided through List::MoreUtils as frontend.

0.418 2017-03-29
    - clarify license conditions

0.417_001 2017-03-22
    - Divorce List::MoreUtils and List::MoreUtils::XS
    - Change license to Apache 2.0 to avoid code stealing without credits
    - Don't support Perl 5.6 out of the box anymore


Makefile.PL
#!perl

use strict;
use warnings;

require 5.008_001;

BEGIN
{
    unless (grep { $_ eq "." } @INC)
    {
        use lib ".";
    }
}

use inc::latest 'Capture::Tiny';
use inc::latest 'Config::AutoConf';

if (inc::latest->can("write"))
{
    inc::latest->write("inc");
    for my $mod (inc::latest->loaded_modules)
    {
        inc::latest->bundle_module($mod, "inc");
    }
}

use ExtUtils::MakeMaker;

use inc::Config::AutoConf::LMU ();

inc::Config::AutoConf::LMU->_set_argv(@ARGV);    # XXX hack because we cannot construct for global use
inc::Config::AutoConf::LMU->check_lmu_prerequisites();
inc::Config::AutoConf::LMU->write_config_h("LMUconfig.h");

if (-d "Sandbox")
{
    eval "use Sandbox::Tumble ();";
    $@ and die $@;
    eval "use File::Path ();";
    File::Path->import;
    -d 't/xs' and rmtree('t/xs');
    Sandbox::Tumble->tumble(qw(t));
}


# Should we build the XS version?
my %RUN_DEPS = (
    'XSLoader'        => "0.22",
);
my %BUNDLE_CONFIGURE_DEPS = (
    'inc::latest'      => '0.500',
    'Config::AutoConf' => '0.315',
);
my %CONFIGURE_DEPS = (
    'Carp'                => 0,
    'ExtUtils::MakeMaker' => 0,
    'File::Basename'      => 0,
    'File::Copy'          => 0,
    'File::Path'          => 0,
    'File::Spec'          => 0,
    'IPC::Cmd'            => 0,
    'base'                => 0,
);
my %BUILD_DEPS = ();

my %TEST_DEPS = (
    'Storable'   => 0,
    'Test::More' => 0.96,
);
my @XS_FILES = qw(XS.xs);

my $conflictMsg = <<EOCM;
***
    This version of %s conflicts with
    the version of module %s (%s) you have installed.

    It's strongly recommended that you update it after
    installing this version of %s.
***
EOCM

my $selfConflictMsg = <<EOSCM;
***
    This version of %s conflicts with
    the version of %s (%s) you have installed.

    It's strongly recommended that you carefully read
    the Changes file and test your application with
    this version before finally upgrading.

    See local::lib how to do it easily.
***
EOSCM

WriteMakefile1(
    META_MERGE => {
        'meta-spec' => {version => 2},
        resources   => {
            homepage   => 'https://metacpan.org/release/List-MoreUtils-XS',
            repository => {
                url  => 'https://github.com/perl5-utils/List-MoreUtils-XS.git',
                web  => 'https://github.com/perl5-utils/List-MoreUtils-XS',
                type => 'git',
            },
            bugtracker => {
                web    => 'https://rt.cpan.org/Dist/Display.html?Name=List-MoreUtils-XS',
                mailto => 'bug-List-MoreUtils-XS@rt.cpan.org',
            },
            license => 'apache_2_0',
        },
        prereqs => {
            develop => {
                requires => {
                    'Test::CPAN::Changes'                 => 0,
                    'Test::CheckManifest'                 => 0,
                    'Module::CPANTS::Analyse'             => '0.96',
                    'Test::Kwalitee'                      => 0,
                    'Test::Pod'                           => 0,
                    'Test::Pod::Coverage'                 => 0,
                    'Test::Pod::Spelling::CommonMistakes' => 0,
                    'Test::Spelling'                      => 0,
                    'Test::WriteVariants'                 => '0.013',
                    %BUNDLE_CONFIGURE_DEPS,
                },
            },
            configure => {
                requires   => {%CONFIGURE_DEPS},
                recommends => {%BUNDLE_CONFIGURE_DEPS},
            },
            build => {requires => {%BUILD_DEPS}},
            test  => {
                requires   => {%TEST_DEPS},
                recommends => {'Test::LeakTrace' => 0}
            },
            runtime => {
                requires => {%RUN_DEPS},
            },
        },
    },
    NAME         => 'List::MoreUtils::XS',
    ABSTRACT     => 'Provide the stuff missing in List::Util in XS',
    VERSION_FROM => 'lib/List/MoreUtils/XS.pm',
    AUTHOR       => [
        'Tassilo von Parseval <tassilo.von.parseval@rwth-aachen.de>',
        'Adam Kennedy <adamk@cpan.org>',
        'Jens Rehsack <rehsack@cpan.org>'
    ],
    LICENSE            => 'apache_2_0',
    CONFIGURE_REQUIRES => \%CONFIGURE_DEPS,
    PREREQ_PM          => \%RUN_DEPS,
    BUILD_REQUIRES     => \%BUILD_DEPS,
    TEST_REQUIRES      => \%TEST_DEPS,
    XS                 => {map { (my $tgt = $_) =~ s/\.xs$/\.c/; $_ => $tgt; } @XS_FILES},
    MAGICXS            => 1,
    INC                => "-I.",
    (-d "Sandbox" ? (realclean => {FILES => "inc/latest* inc/inc_* t/xs"}) : ()),
    depend => {'$(FIRST_MAKEFILE)' => '$(VERSION_FROM)'},
    test   => {TESTS               => join(' ', 't/*.t', 't/xs/*.t', 'xt/*.t')},
    # Otherwise 'cxinc' isn't defined
    ($] < 5.012 ? (DEFINE => '-DPERL_EXT') : (),),
);

sub CheckConflicts
{
    my %params    = @_;
    my %conflicts = %{ $params{CONFLICTS} };
    my $found     = 0;

    while (my ($module, $version) = each(%conflicts))
    {
        undef $@;
        eval "require $module";
        next if $@;
        my $installed = eval "\$" . $module . "::VERSION";
        if ($installed le $version)
        {
            ++$found;
            my $warning =
              (0 == index($module, $params{NAME}, 0))
              ? sprintf($selfConflictMsg, $params{NAME}, $module, $installed)
              : sprintf($conflictMsg, $params{NAME}, $module, $installed, $params{NAME});
            warn $warning;
        }
    }

    return !$found;
}

sub WriteMakefile1
{    # originally written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
    my %params       = @_;
    my $eumm_version = $ExtUtils::MakeMaker::VERSION;
    $eumm_version = eval $eumm_version;
    die "EXTRA_META is deprecated" if (exists($params{EXTRA_META}));
    die "License not specified"    if (!exists($params{LICENSE}));
    $params{TEST_REQUIRES}
      and $eumm_version < 6.6303
      and $params{BUILD_REQUIRES} = {%{$params{BUILD_REQUIRES} || {}}, %{delete $params{TEST_REQUIRES}}};
    #EUMM 6.5502 has problems with BUILD_REQUIRES
    $params{BUILD_REQUIRES}
      and $eumm_version < 6.5503
      and $params{PREREQ_PM} = {%{$params{PREREQ_PM} || {}}, %{delete $params{BUILD_REQUIRES}}};
    ref $params{AUTHOR}
      and "ARRAY" eq ref $params{AUTHOR}
      and $eumm_version < 6.5702
      and $params{AUTHOR} = join(", ", @{$params{AUTHOR}});
    delete $params{CONFIGURE_REQUIRES}    if ($eumm_version < 6.52);
    delete $params{MIN_PERL_VERSION}      if ($eumm_version < 6.48);
    delete $params{META_MERGE}            if ($eumm_version < 6.46);
    delete $params{META_ADD}{prereqs}     if ($eumm_version < 6.58);
    delete $params{META_ADD}{'meta-spec'} if ($eumm_version < 6.58);
    delete $params{META_ADD}              if ($eumm_version < 6.46);
    delete $params{LICENSE}               if ($eumm_version < 6.31);
    delete $params{AUTHOR}                if ($] < 5.005);
    delete $params{ABSTRACT_FROM}         if ($] < 5.005);
    delete $params{BINARY_LOCATION}       if ($] < 5.005);

    # more or less taken from Moose' Makefile.PL
    if ($params{CONFLICTS})
    {
        my $ok = CheckConflicts(%params);
        exit(0) if ($params{PREREQ_FATAL} and not $ok);
        my $cpan_smoker = grep { $_ =~ m/(?:CR_SMOKER|CPAN_REPORTER|AUTOMATED_TESTING)/ } keys %ENV;
        unless ($cpan_smoker || $ENV{PERL_MM_USE_DEFAULT})
        {
            sleep 4 unless ($ok);
        }
        delete $params{CONFLICTS};
    }

    WriteMakefile(%params);
}


Sandbox/Tumble.pm
package Sandbox::Tumble;

use strict;
use warnings;

use Cwd qw();
use File::Spec qw();
use Test::WriteVariants 0.013;

use FindBin qw();

$| = 1;

sub tumble
{
    my ($class, $output_dir) = @_;

    my $template_dir = Cwd::abs_path(File::Spec->catdir($FindBin::RealBin, "t", "inline"));
    my $test_writer  = Test::WriteVariants->new();
    $test_writer->allow_dir_overwrite(1);
    $test_writer->allow_file_overwrite(1);

    $test_writer->write_test_variants(
        input_tests => $test_writer->find_input_inline_tests(
            search_patterns => ["*.pm"],
            ,
            search_dirs => [$template_dir],
        ),
        variant_providers => ["LMU::TestVariants",],
        output_dir        => $output_dir,
    );
}

package LMU::TestVariants::Dist;

use strict;
use warnings;

sub provider
{
    my ($self, $path, $context, $tests, $variants) = @_;
    my $mod_lmu  = $context->new_module_use('List::MoreUtils::XS' => [':all']);
    my $mod_ctx  = $context->new_module_use(lib                   => [File::Spec->catdir(qw(t lib))]);
    my $strict   = $context->new_module_use(strict                => [qw(subs vars refs)]);
    my $warnings = $context->new_module_use(warnings              => ['all']);

    # statically generate both at dist authoring stage and decide about tests to run at configure stage
    $variants->{xs} = $context->new($mod_lmu, $mod_ctx, $warnings, $strict,);
}

1;


lib/List/MoreUtils/XS.pm
package List::MoreUtils::XS;

use 5.008_001;
use strict;
use warnings;
use base ('Exporter');

use vars qw{$VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS};

$VERSION = '0.429_002';

@EXPORT    = ();
@EXPORT_OK = qw(any all none notall one
  any_u all_u none_u notall_u one_u
  reduce_u reduce_0 reduce_1
  true false
  insert_after insert_after_string
  apply indexes slide
  after after_incl before before_incl
  firstidx lastidx onlyidx
  firstval lastval onlyval
  firstres lastres onlyres
  singleton duplicates frequency occurrences mode
  each_array each_arrayref
  pairwise natatime slideatatime
  arrayify mesh zip6 uniq listcmp
  samples minmax minmaxstr part
  bsearch bsearchidx binsert bremove lower_bound upper_bound equal_range
  qsort);
%EXPORT_TAGS = (all => \@EXPORT_OK);

# Load the XS at compile-time so that redefinition warnings will be
# thrown correctly if the XS versions of part or indexes loaded

# PERL_DL_NONLAZY must be false, or any errors in loading will just
# cause the perl code to be tested
local $ENV{PERL_DL_NONLAZY} = 0 if $ENV{PERL_DL_NONLAZY};

use XSLoader ();
XSLoader::load("List::MoreUtils::XS", "$VERSION");

=pod

=head1 NAME

List::MoreUtils::XS - Provide compiled List::MoreUtils functions

=head1 SYNOPSIS

  use List::MoreUtils::XS (); # doesn't export anything
  use List::MoreUtils ':all'; # required to import functions

  my @procs = get_process_stats->fetchall_array;
  # sort by ppid, then pid
  qsort { $a->[3] <=> $b->[3] or $a->[2] <=> $b->[2] } @procs;
  while( @procs ) {
      my $proc = shift @procs;
      my @children = equal_range { $_->[3] <=> $proc->[2] } @procs;
  }

  my @left = qw(this is a test);
  my @right = qw(this is also a test);
  my %rlinfo = listcmp @left, @right;

  # on unsorted
  my $i = firstidx { $_ eq 'yeah' } @foo;
  # on sorted - always first, but might not be 'yeah'
  my $j = lower_bound { $_ cmp 'yeah' } @bar;
  # on sorted - any of occurrences, is surely 'yeah'
  my $k = bsearchidx { $_ cmp 'yeah' } @bar;

=head1 DESCRIPTION

List::MoreUtils::XS is a backend for List::MoreUtils. Even if it's possible
(because of user wishes) to have it practically independent from
L<List::MoreUtils>, it technically depend on C<List::MoreUtils>. Since it's
only a backend, the API is not public and can change without any warning.

=head1 SEE ALSO

L<List::Util>, L<List::AllUtils>

=head1 AUTHOR

Jens Rehsack E<lt>rehsack AT cpan.orgE<gt>

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

Tassilo von Parseval E<lt>tassilo.von.parseval@rwth-aachen.deE<gt>

=head1 COPYRIGHT AND LICENSE

Some parts copyright 2011 Aaron Crane.

Copyright 2004 - 2010 by Tassilo von Parseval

Copyright 2013 - 2017 by Jens Rehsack

All code added with 0.417 or later is licensed under the Apache License,
Version 2.0 (the "License"); you may not use this file except in compliance
with the License. You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

All code until 0.416 is licensed under the same terms as Perl itself,
either Perl version 5.8.4 or, at your option, any later version of
Perl 5 you may have available.

=cut

1;
